### ISSUE TITLE
Changing Colorscheme causes highlight settings in diagnostics  to be cleared

### ISSUE THREAD

**Issue by tjdevries** [neovim/neovim#12579]
Maybe we should think a little more about the call for this.
Changing color scheme causes the highlight setting in diagnostics to be cleared.

_Originally posted by @h-michael in https://github.com/neovim/neovim/pull/12333_

**Comment by dimascyriaco** [comment]
Is there a workaround for this problem?

**Comment by tjdevries** [comment]
Hmmm, perhaps if we called `vim.lsp.diagnostic._define_default_signs_and_highlights()` in plugin somewhere, then it would work?

@dimascyriaco Can you detail what you do to experience something not working? I can try and fix at some point with some more recent repro steps.

**Comment by dimascyriaco** [comment]
@tjdevries I think I got a reproducible minimal `init.lua`:

```lua
vim.cmd [[packadd packer.nvim]]

require('packer').startup(function()
  use {'wbthomason/packer.nvim', opt = true}
  use { 'neovim/nvim-lspconfig' }
  use { 'arcticicestudio/nord-vim' }
end)

require'lspconfig'.tsserver.setup{}

vim.cmd 'colorscheme nord'
```

Running this with `nvim --clean -u init.lua --cmd "set rtp+=$(pwd)" --cmd "set packpath+=$(pwd)" init.lua` (don't know if this is the best way to do it). I cloned those plugins direct into this folder (inside `./pack/plugin-name/start`).

If I comment `vim.cmd 'colorscheme nord'` the highlights work.

Example:

![Screen Shot 2021-03-14 at 14 52 27](https://user-images.githubusercontent.com/65256/111078761-dfa58800-84d5-11eb-979c-d6415a3b8eb2.png)

vs

![Screen Shot 2021-03-14 at 14 51 35](https://user-images.githubusercontent.com/65256/111078765-e7fdc300-84d5-11eb-8eb1-890af9eef56b.png)

**Comment by mjlbach** [comment]
Related: https://github.com/neovim/neovim/issues/14034

@clason This is what we were talking about on IRC with @bfredl. Nord, as with many colorschemes, call `hi clear` on initialization which clears our highlight groups for underline etc. If you coment out `hi clear` [here](https://github.com/arcticicestudio/nord-vim/blob/ea7ff9c343392ec6dfac4e9ec3fe0c45afb92a40/colors/nord.vim#L9) the problem is resolved.

**Comment by erw7** [comment]
> Nord, as with many colorschemes, call hi clear on initialization which clears our highlight groups for underline etc.

Therefore, if users want to set up their own highlight, I think they need to do so in the form of `autocmd ColorScheme * hi ...` before setting up the colorscheme.

In the case of the @dimas-cyriaco configuration, I think the following will work.

```vim
vim.cmd [[packadd packer.nvim]]

require('packer').startup(function()
  use { 'wbthomason/packer.nvim', opt = true }
  use { 'neovim/nvim-lspconfig' }
  use { 'arcticicestudio/nord-vim' }
end)

require'lspconfig'.tsserver.setup{}

autocmd ColorScheme * call v:lua.vim.lsp.diagnostic._define_default_signs_and_highlights()

vim.cmd 'colorscheme nord'
```

We may be able to configure `autocmd` settings in `_define_default_signs_and_highlights()`, but that will override the user's settings if called in the following order.

```vim
" User highlighting settings
autocmd ColorScheme * hi LspDiagnositcsSignError …
" _define_default_signs_and_highlights() call
lua require'lspconfig.tsserver‘.setup{}
colorscheme nord
```

In the end, problems caused by the order of calls need to be considered by the users themselves, and I think the only way to do that is to document them.

**Comment by mjlbach** [comment]
I made a doc PR using the above snippet, I agree that adding an autocommand in _define_default_signs_and_highlights() seems like overkill when most colorschemes will define these anyways. We could have the autocommand call a function which checks if the highlights are defined before overriding them, but that seems like unnecessary complexity.

**Comment by clason** [comment]
@mjlbach But that's not the only issue (I meant to add a comment on my own issue):

1. neovim (differently from vim) enables `syntax on` by default -- but it only does so _after_ `init.vim` is processed, and
> ":syntax on" overrules existing colors, ":syntax enable" only sets groups that weren't set yet. 

2. https://github.com/neovim/neovim/pull/14060 only addresses restoring default links, not groups that are explicitly set via `guifg=`etc.

Point 1 is only relevant to setting color schemes in your `init.vim`, of course: adding an explicit `syn on` at the start and setting `colorscheme` _before_ requiring `lspconfig` also sets up the highlight groups correctly.

Point 2, on the other hand, may help with changing colorschemes while running.

**Comment by clason** [comment]
I closed #14034 as it is in fact a duplicate of this issue (doing my part to get the issue count down!), but I think the underlying issues can and should be addressed to improve the situation:

1. either move the implicit `syn on` _before_ processing user config or, quoting @bfredl from Gitter,
> source syntax/syncolor.vim before the users init.vim runs (essentially, the vim splits it default colors in two sets (basic ui hl, and syntax hl))
2. extending the patch in #14060 to also cover non-linked highlight groups

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/14771 was merged, and also "syntax on" and "syntax enable" are now equivalent.
